### PR TITLE
CEL wiring in rule resolver

### PR DESCRIPTION
**Pull request by aramase**

### CODE REVIEW

File: pkg/registry/rbac/cel/matcher.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+type CELMatcher struct {
+	CompilationResults []CompilationResult
+
+	// These track if any expressions use fieldSelector and labelSelector,
+	// so construction of data passed to the CEL expression can be optimized if those fields are unused.
+	UsesLabelSelector bool
+	UsesFieldSelector bool
+}
+
+func (c *CELMatcher) Eval(ctx context.Context, conditionalAttributes ConditionalAttributes) (bool, error) {
```
Comment (enj) [inline pkg/registry/rbac/cel/matcher.go:35]: Just to be cautious: return `false, nil` early if `len(CompilationResults) == 0 `

File: pkg/registry/rbac/cel/compile.go [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+	return apiservercel.NewObjectType("kubernetes.LabelSelectorAttributes", fields(
+		field("requirements", apiservercel.NewListType(buildSelectorRequirementType(field, fields), -1), false),
+	))
+}
+
+func buildSelectorRequirementType(field func(name string, declType *apiservercel.DeclType, required bool) *apiservercel.DeclField, fields func(fields ...*apiservercel.DeclField) map[string]*apiservercel.DeclField) *apiservercel.DeclType {
+	return apiservercel.NewObjectType("kubernetes.SelectorRequirement", fields(
+		field("key", apiservercel.StringType, false),
+		field("operator", apiservercel.StringType, false),
+		field("values", apiservercel.NewListType(apiservercel.StringType, -1), false),
```
Comment (enj) [inline pkg/registry/rbac/cel/compile.go]: nit order should always be:

1. user
2. uid
3. groups
4. extra

File: pkg/registry/rbac/validation/rule_test.go [hunk 3]
```diff
@@ -275,3 +283,121 @@ func TestAppliesTo(t *testing.T) {
...
+			},
+			expected: true,
+		},
+		{
+			name: "node operation restricted based on field selector",
+			conditionalClusterRoleBinding: &rbacv1alpha1.ConditionalClusterRoleBinding{
+				Conditions: []rbacv1alpha1.Condition{
+					{
+						// determine it's a node by checking the prefix and asserting the name is not just the prefix
+						Expression: "request.user.startsWith('system:node') && request.user != 'system:node'",
```
Comment (enj) [inline pkg/registry/rbac/validation/rule_test.go]: btw I think `!=` violates the no negative rule?

```suggestion
						Expression: "request.user.startsWith('system:node:') && request.user != 'system:node:'",
```

File: pkg/registry/rbac/validation/rule_test.go [hunk 4]
```diff
@@ -275,3 +283,121 @@ func TestAppliesTo(t *testing.T) {
...
+					},
+					{
+						// it needs to be in system:nodes group
+						Expression: "request.groups.exists(g, g == 'system:nodes')",
+					},
+					{
+						// now match the field selector to restrict to a specific node
+						Expression: "request.resourceAttributes.fieldSelector.requirements.exists(r, r.key == 'spec.nodeName' && sets.equivalent(r.values, [request.user.substring(12)]))",
```
Comment (enj) [inline pkg/registry/rbac/validation/rule_test.go]: missing operator.

**Review by enj — CHANGES_REQUESTED** [review]
Minor comments and some repeats from yesterday.

File: pkg/registry/rbac/validation/rule_test.go (hunk 3 above)
Comment (aramase) [reply pkg/registry/rbac/validation/rule_test.go]: changed it to check the length instead

File: pkg/registry/rbac/validation/rule_test.go [hunk 5]
```diff
@@ -275,3 +283,121 @@ func TestAppliesTo(t *testing.T) {
...
+						// determine it's a node by checking the prefix and asserting the name is not just the prefix
+						Expression: "request.user.startsWith('system:node:') && size(request.user) > size('system:node:')",
+					},
+					{
+						// it needs to be in system:nodes group
+						Expression: "request.groups.exists(g, g == 'system:nodes')",
+					},
+					{
+						// now match the field selector to restrict to a specific node
+						Expression: "request.resourceAttributes.fieldSelector.requirements.exists(r, r.key == 'spec.nodeName' && r.operator == '=' && sets.equivalent(r.values, [request.user.substring(12)]))",
```
Comment (enj) [inline pkg/registry/rbac/validation/rule_test.go]:
```suggestion
						Expression: "request.resourceAttributes.fieldSelector.requirements.exists(r, r.key == 'spec.nodeName' && r.operator == '=' && sets.equivalent(r.values, [request.user.substring(size('system:node:'))]))",
```

**Pull request merged**